vault/expiration.go

```diff
@@ -1806,6 +1823,55 @@ func (m *ExpirationManager) deleteLockForLease(id string) {
 	m.lockPerLease.Delete(id)
 }
 
+func (m *ExpirationManager) sendEventByID(eventType logical.EventType, leaseID string) {
+	le, err := m.loadEntry(context.Background(), leaseID)
+	if err != nil {
+		m.logger.Warn("Error fetching lease information", "error", err)
+		return
+	}
+	m.sendEvent(eventType, le)
+}
+
+func (m *ExpirationManager) sendEvent(eventType logical.EventType, le *leaseEntry) {
+	data := &logical.EventData{}
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		m.logger.Warn("Error creating uuid", "error", err)
+		return
+	}
+	ns := namespace.RootNamespace
+	if le.namespace != nil {
+		ns = le.namespace
+	}
+	data.Id = id
+	data.EntityIds = []string{le.ClientToken}
+	structMap := map[string]interface{}{
+		"path":              le.Path,
+		"expire_time":       le.ExpireTime.Format(time.RFC3339),
+		"last_renewal_time": le.LastRenewalTime.Format(time.RFC3339),
+		"version":           strconv.Itoa(le.Version),
+		"login_role":        le.LoginRole,
+		"client_token_type": le.ClientTokenType.String(),
+		"namespace":         ns.Path,
+	}
+	if le.Auth != nil {
```

Comment:
My recollection is that in a leaseEntry, we commonly only preserve a handful of Auth fields: namespace, policies, and auth method.  We need the Auth structure to route renewals IIRC, and for some metrics we emit, but we try to minimize how much of it we keep around, for memory reasons.

Reply:
I understanding of this code is far from complete, but it looks like the only that `le.Auth` is trimmed is for in-memory storage in `m.irrevocable`, `m.nonexpiring`, and `m.pending`, but it looks like it is always put into storage with all the fields?

Reply:
(And so far, we are only using the `le.Auth` that comes from storage in these code paths.)

Reply:
Hmm, you're right it looks like, I'm probably thinking of a big PR I never merged.  There are exceptions, like ControlGroupStore.createToken and wrapInCubbyHole, but in the main what you say is true as far as I can tell.